Python/Product/PythonTools/PythonTools/Editor/PythonEditorServices.cs

```diff
@@ -0,0 +1,122 @@
+﻿// Python Tools for Visual Studio
+// Copyright(c) Microsoft Corporation
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the License); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
+// OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using System;
+using System.ComponentModel.Composition;
+using Microsoft.PythonTools.Intellisense;
+using Microsoft.VisualStudio.ComponentModelHost;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Classification;
+using Microsoft.VisualStudio.Utilities;
+using Microsoft.VisualStudioTools;
+
+namespace Microsoft.PythonTools.Editor {
+    /// <summary>
+    /// Provides centralized access to services used by the editor.
+    /// </summary>
+    /// <remarks>
+    /// MEF services provided by VS should be imported eagerly using the
+    /// [Import] attribute on a public field.
+    /// 
+    /// Traditional services should be lazily imported on first use, and
+    /// uses should be audited to ensure they all occur on the UI thread.
+    /// 
+    /// Services provided by PTVS should be lazily loaded on first access.
+    /// Otherwise, we may end up with circular imports in MEF composition.
+    /// </remarks>
+    [Export]
+    sealed class PythonEditorServices {
+        [ImportingConstructor]
+        public PythonEditorServices([Import(typeof(SVsServiceProvider))] IServiceProvider site) {
+            Site = site;
+            ComponentModel = Site.GetComponentModel();
+            _errorTaskProvider = new Lazy<ErrorTaskProvider>(CreateTaskProvider<ErrorTaskProvider>);
+            _commentTaskProvider = new Lazy<CommentTaskProvider>(CreateTaskProvider<CommentTaskProvider>);
+            _unresolvedImportSquiggleProvider = new Lazy<UnresolvedImportSquiggleProvider>(CreateImportSquiggleProvider);
+            _analysisEntryService = new Lazy<AnalysisEntryService>(() => ComponentModel.GetService<AnalysisEntryService>());
+        }
+
+        public readonly IServiceProvider Site;
+
+        #region PythonToolsService
+
+        private PythonToolsService _python;
+
+        internal void SetPythonToolsService(PythonToolsService service) {
```

Comment:
It's hard to guarantee that the GetService will be on the UI thread for this particular one, and definitely preferable to have it be set in the constructor of the service. We may need this class when documents are opened before our package is loaded (because a chain of MEF imports can get us there), and if any of those get the service then we'll trigger the load via TryGetPythonToolsService. Not entirely sure why I wrote it this way, but it keeps the property looking neat.